# **Module 4: Keras Assignment**


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import tensorflow as tf
from tensorflow import keras
import warnings
warnings.filterwarnings('ignore')

In [2]:
df=pd.read_csv('diabetes.csv')
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI   
0            6      148             72             35        0  33.6  \
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

**TASK**

**A. Build a sequential model using Keras on top of this Diabetes dataset to find out if the
patient has diabetes or not, using ‘Pregnancies’, ‘Glucose’ & ‘BloodPressure’ as
independent columns.
a. This model should have 1 hidden layer with 8 nodes
b. Use Stochastic Gradient as the optimization algorithm
c. Fit the model, with number of epochs to be 100 and batch size to be 10**

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [4]:
df.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

NO NULL Values so will model directly

In [5]:
x=df.iloc[:,:3]
x

Pregnancies  Glucose  BloodPressure
0              6      148             72
1              1       85             66
2              8      183             64
3              1       89             66
4              0      137             40
..           ...      ...            ...
763           10      101             76
764            2      122             70
765            5      121             72
766            1      126             60
767            1       93             70

[768 rows x 3 columns]

In [6]:
y=df['Outcome']
y

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64

In [7]:
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
x_scaled=sc.fit_transform(x)
x_scaled[1:10]

array([[-0.84488505, -1.12339636, -0.16054575],
       [ 1.23388019,  1.94372388, -0.26394125],
       [-0.84488505, -0.99820778, -0.16054575],
       [-1.14185152,  0.5040552 , -1.50468724],
       [ 0.3429808 , -0.15318486,  0.25303625],
       [-0.25095213, -1.34247638, -0.98770975],
       [ 1.82781311, -0.184482  , -3.57259724],
       [-0.54791859,  2.38188392,  0.04624525],
       [ 1.23388019,  0.12848945,  1.39038675]])

In [8]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x_scaled,y,test_size=0.2)

In [9]:
x_train.shape

(614, 3)

In [11]:
from tensorflow.keras.optimizers import SGD
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential


# create model
model = Sequential()
model.add(Dense(8, input_shape=(3,), activation='relu'))
model.add(Dense(1, activation='sigmoid'))
# Compile model
epochs = 100
learning_rate = 0.1
decay_rate = learning_rate / epochs
momentum = 0.8
sgd = SGD(learning_rate=learning_rate, momentum=momentum, decay=decay_rate, nesterov=False)
model.compile(loss='binary_crossentropy', optimizer=sgd, metrics=['accuracy'])
# Fit the model
model.fit(x_train, y_train, validation_split=0.33, epochs=epochs, batch_size=10, verbose=2)

Epoch 1/100
42/42 - 5s - loss: 0.5963 - accuracy: 0.6399 - val_loss: 0.4846 - val_accuracy: 0.7931 - 5s/epoch - 110ms/step
Epoch 2/100
42/42 - 0s - loss: 0.5284 - accuracy: 0.7445 - val_loss: 0.4897 - val_accuracy: 0.7537 - 204ms/epoch - 5ms/step
Epoch 3/100
42/42 - 0s - loss: 0.5243 - accuracy: 0.7251 - val_loss: 0.4624 - val_accuracy: 0.7635 - 193ms/epoch - 5ms/step
Epoch 4/100
42/42 - 0s - loss: 0.5371 - accuracy: 0.7105 - val_loss: 0.4711 - val_accuracy: 0.7833 - 191ms/epoch - 5ms/step
Epoch 5/100
42/42 - 0s - loss: 0.5266 - accuracy: 0.7105 - val_loss: 0.4670 - val_accuracy: 0.8128 - 188ms/epoch - 4ms/step
Epoch 6/100
42/42 - 0s - loss: 0.5266 - accuracy: 0.7275 - val_loss: 0.4727 - val_accuracy: 0.8276 - 208ms/epoch - 5ms/step
Epoch 7/100
42/42 - 0s - loss: 0.5243 - accuracy: 0.7226 - val_loss: 0.4567 - val_accuracy: 0.7882 - 203ms/epoch - 5ms/step
Epoch 8/100
42/42 - 0s - loss: 0.5263 - accuracy: 0.7226 - val_loss: 0.4673 - val_accuracy: 0.8030 - 196ms/epoch - 5ms/step
Epoch 9/1

In [12]:
from sklearn.metrics import confusion_matrix , classification_report
import numpy as np
y_pred = model.predict(x_test)
y_pred_classes = [np.argmax(element) for element in y_pred]

print("Classification Report: \n", classification_report(y_test, y_pred_classes))

5/5 [==============================] - 0s 5ms/step
Classification Report: 
               precision    recall  f1-score   support

           0       0.59      1.00      0.74        91
           1       0.00      0.00      0.00        63

    accuracy                           0.59       154
   macro avg       0.30      0.50      0.37       154
weighted avg       0.35      0.59      0.44       154



**B. Build another sequential model where ‘Outcome’ is the dependent variable and all
other columns are predictors.
a. This model should have 3 hidden layers with 16 nodes in each layer
b. Use ‘adam’ as the optimization algorithm
c. Fit the model, with number of epochs to be 150 and batch size to be 10**

In [16]:
predictors=df.iloc[:,:8]
predictors

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI   
0              6      148             72             35        0  33.6  \
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
1                       0.351   31  
2                       0.672   32  
3                       0.167   21  
4                       2.288   33  
..                        ...  ...  
763                     0.171   63  
764                     0.340   27  
765                     0.245   30  
766                     0.349   47  
767                     0.315   23  

[768 rows x 8 columns]

In [17]:
target=df['Outcome']
target

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64

In [19]:
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
predictors_scaled=sc.fit_transform(x)
predictors_scaled[1:10]

array([[-0.84488505, -1.12339636, -0.16054575,  0.53090156, -0.69289057,
        -0.68442195, -0.36506078, -0.19067191],
       [ 1.23388019,  1.94372388, -0.26394125, -1.28821221, -0.69289057,
        -1.10325546,  0.60439732, -0.10558415],
       [-0.84488505, -0.99820778, -0.16054575,  0.15453319,  0.12330164,
        -0.49404308, -0.92076261, -1.04154944],
       [-1.14185152,  0.5040552 , -1.50468724,  0.90726993,  0.76583594,
         1.4097456 ,  5.4849091 , -0.0204964 ],
       [ 0.3429808 , -0.15318486,  0.25303625, -1.28821221, -0.69289057,
        -0.81134119, -0.81807858, -0.27575966],
       [-0.25095213, -1.34247638, -0.98770975,  0.71908574,  0.07120427,
        -0.12597727, -0.676133  , -0.61611067],
       [ 1.82781311, -0.184482  , -3.57259724, -1.28821221, -0.69289057,
         0.41977549, -1.02042653, -0.36084741],
       [-0.54791859,  2.38188392,  0.04624525,  1.53455054,  4.02192191,
        -0.18943689, -0.94794368,  1.68125866],
       [ 1.23388019,  0.12848945

In [20]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(predictors_scaled,target,test_size=0.2)

In [21]:
x_train.shape

(614, 8)

** ‘adam’ as the optimization algorithm **

In [22]:
from tensorflow import keras
from tensorflow.keras import layers

model = keras.Sequential([
    layers.Dense(16, activation='relu', input_shape=[8]),
    layers.Dense(16, activation='relu'),
    layers.Dense(16, activation='relu'),
    layers.Dense(1),
])
model.compile(
    optimizer='adam',
    loss='mse',
     metrics=['accuracy'])

model.fit(x_train, y_train, validation_split=0.33, epochs=150, batch_size=10, verbose=2)

from sklearn.metrics import confusion_matrix , classification_report
import numpy as np
y_pred = model.predict(x_test)
y_pred_classes = [np.argmax(element) for element in y_pred]

print("Classification Report: \n", classification_report(y_test, y_pred_classes))

Epoch 1/150
42/42 - 2s - loss: 0.2040 - accuracy: 0.7080 - val_loss: 0.2408 - val_accuracy: 0.6749 - 2s/epoch - 38ms/step
Epoch 2/150
42/42 - 0s - loss: 0.1762 - accuracy: 0.7397 - val_loss: 0.2196 - val_accuracy: 0.6847 - 196ms/epoch - 5ms/step
Epoch 3/150
42/42 - 0s - loss: 0.1659 - accuracy: 0.7543 - val_loss: 0.2123 - val_accuracy: 0.6847 - 203ms/epoch - 5ms/step
Epoch 4/150
42/42 - 0s - loss: 0.1596 - accuracy: 0.7616 - val_loss: 0.2099 - val_accuracy: 0.7044 - 197ms/epoch - 5ms/step
Epoch 5/150
42/42 - 0s - loss: 0.1550 - accuracy: 0.7786 - val_loss: 0.2050 - val_accuracy: 0.6995 - 197ms/epoch - 5ms/step
Epoch 6/150
42/42 - 0s - loss: 0.1503 - accuracy: 0.7908 - val_loss: 0.2048 - val_accuracy: 0.7241 - 197ms/epoch - 5ms/step
Epoch 7/150
42/42 - 0s - loss: 0.1473 - accuracy: 0.7908 - val_loss: 0.1989 - val_accuracy: 0.7291 - 198ms/epoch - 5ms/step
Epoch 8/150
42/42 - 0s - loss: 0.1461 - accuracy: 0.7908 - val_loss: 0.1979 - val_accuracy: 0.7241 - 195ms/epoch - 5ms/step
Epoch 9/15

**CONCLUSION**

By using staochastic gradient optimizer the train accuracy obtained is ~75.6% and the test accuracy obtained is ~79%

By using ADAM optimizer the with the all the features taken into consideration we can observed that after 150epochs the train accuracy was higher around 98% ,However the val_accuracy reduced to 67%.

We can conclude that by uing ADAM with  epochs 150 our model trying to Overfitted.